### PR TITLE
Implement partial transactions for AWS Lambda

### LINKED ISSUE #2999: [META 753] Support Partial Transaction for AWS Lambda
See meta issue for the description and details:
- Meta issue: https://github.com/elastic/apm/issues/753

### PR DESCRIPTION
**Pull request by JonasKunz**
## What does this PR do?

Closes #2999 .

## Checklist

- [x] This is an enhancement of existing features, or a new feature in existing plugins
  - [x] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/main/CHANGELOG.asciidoc)
  - [x] I have added tests that prove my fix is effective or that my feature works
  - [x] ~Added an API method or config option? Document in which version this will be introduced~
  - [x] ~I have made corresponding changes to the documentation~

### CODE REVIEW

File: apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/Transaction.java [hunk 1]
```diff
@@ -179,6 +182,23 @@ public Transaction startNoop() {
         return this;
     }
 
+    @Override
+    public Transaction activate() {
+        if (!wasActivated.get()) {
+            if (!wasActivated.getAndSet(true)) {
+                onFirstActivation();
+            }
+        }
```
Comment (SylvainJuge) [inline apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/Transaction.java:191]: [minor] the outer `if` seems redundant here

File: apm-agent-core/src/test/java/co/elastic/apm/agent/report/PartialTransactionTest.java [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+    @Test
+    public void testNoMoreReportingAfter4xx() {
+        apmServer.stubFor(post(urlEqualTo("/register/transaction")).willReturn(aResponse().withStatus(404)));
+        ServerlessConfiguration serverlessConfig = spyConfig.getConfig(ServerlessConfiguration.class);
+        doReturn(true).when(serverlessConfig).runsOnAwsLambda();
+
+        Transaction tx1 = tracer
+            .startRootTransaction(null)
+            .withName("faas-transaction");
+        tx1.getFaas().withExecution("foo-bar-id");
```
Comment (SylvainJuge) [inline apm-agent-core/src/test/java/co/elastic/apm/agent/report/PartialTransactionTest.java:179]: [question] if I understand correctly the `withExecution` call makes the transaction as a FAAS transaction, in opposition to a "regular" transaction where this is not called.

File: apm-agent-core/src/test/java/co/elastic/apm/agent/MockReporter.java [hunk 3]
```diff
@@ -190,6 +195,23 @@ public void setImmediateRecycling(boolean enable) {
...
+            partialTransactions.add(transaction);
+        }
+        if (partialTransactionHandler != null) {
+            partialTransactionHandler.accept(transaction);
+        }
+    }
+
+    public void setPartialTransactionHandler(Consumer<Transaction> transactionHandler) {
+        this.partialTransactionHandler = transactionHandler;
+    }
```
Comment (SylvainJuge) [inline apm-agent-core/src/test/java/co/elastic/apm/agent/MockReporter.java:213]: [minor] seems unused.

**Review by SylvainJuge — APPROVED** [review]

File: apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/Transaction.java (hunk 1 above)
Comment (JonasKunz) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/Transaction.java:191]: This is a small performance optimization: `get()` on an atomic is a volatile read, which costs the same as a normal read. `getAndSet()` is an atomic operation (similar to CAS) and therefore comes with the memory-barrier penalty like other atomic operations. I'll add a comment to clarify

File: apm-agent-core/src/test/java/co/elastic/apm/agent/report/PartialTransactionTest.java (hunk 2 above)
Comment (JonasKunz) [reply apm-agent-core/src/test/java/co/elastic/apm/agent/report/PartialTransactionTest.java:179]: Exactly.

File: apm-agent-core/src/test/java/co/elastic/apm/agent/MockReporter.java (hunk 3 above)
Comment (JonasKunz) [reply apm-agent-core/src/test/java/co/elastic/apm/agent/MockReporter.java:213]: It is actually used by the `PartialTransactionTest` to plugin a real `PartialTransactionReporter` into the `MockReporter`.

**Pull request merged**